src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -512,11 +760,15 @@ def frobenius_norm(self):
             return self._frobenius_norm
         K = self.base_over_constants_field()
         n = K.degree(self._Fq)
-        char = self.characteristic()
+        r = self.rank()
+        p = self.characteristic()
         norm = K(self.coefficients()[-1]).norm()
-        self._frobenius_norm = ((-1)**n)*(char**(n/char.degree())) / norm
+        self._frobenius_norm = (-1) ** (n*r - n - r) \
+                               * norm**(-1) \
+                               * p ** (n//p.degree())
         return self._frobenius_norm
 
+    @cached_method
     def frobenius_trace(self):
```

Comment:
In this method, you use the crystalline matrix in all cases. I think it is okay but aren't there any case where the motive matrix is faster to compute?

Reply:
Yup. Will do take care of that.

Reply:
Any objection to making `DrinfeldModuleMorphism._motive_matrix` public? It's interesting in its own right and would be useful in our case.

Reply:
What do you mean by public? Without the initial underscore?

Reply:
Yup.

Reply:
Well, eventually, I plan to have a method `motive` to create the motive of a Drinfeld module or a morphism between those, and a method `matrix` to get the matrix of a morphism of Anderson motives.
So, if you make the method `_motive_matrix` public now, it means that it will need to be deprecated afterwards. Therefore, I prefer keeping it as it is.